tests/test_protocol.py

```diff
@@ -0,0 +1,152 @@
+        ("pt:optional[pt:int]", "conint(ge=Int32.min(), le=Int32.max()) | None"),
+        ("pt:dict[pt:str, pt:int]", "dict[str, conint(ge=Int32.min(), le=Int32.max())]"),
+        (
+            "pt:list[pt:union[pt:dict[pt:str, pt:int], pt:list[pt:bytes]]]",
+            "tuple[dict[str, conint(ge=Int32.min(), le=Int32.max())] | tuple[bytes]]",
+        ),
+        (
+            "pt:optional[pt:dict[pt:union[pt:str, pt:int], pt:list[pt:union[pt:float, pt:bool]]]]",
+            "dict[str | conint(ge=Int32.min(), le=Int32.max()), tuple[confloat(ge=Double.min(), le=Double.max()) | bool]] | None",  # noqa: E501
+        ),
+    ],
+)
+def test_parse_performative_annotation(annotation: str, expected: str):
+    """Test parse_performative_annotation."""
+    assert performatives.parse_annotation(annotation) == expected
+
+
+@pytest.mark.parametrize(
+    "protocol_spec",
+    [
+        PROTOCOL_FILES["balances.yaml"],
+        # PROTOCOL_FILES["bridge.yaml"],  # noqa: ERA001
+        # PROTOCOL_FILES["cross_chain_arbtrage.yaml"],  # noqa: ERA001
+        # PROTOCOL_FILES["default.yaml"], # noqa: ERA001
+        PROTOCOL_FILES["liquidity_provision.yaml"],
+        PROTOCOL_FILES["markets.yaml"],
+        PROTOCOL_FILES["ohlcv.yaml"],
+        # PROTOCOL_FILES["order_book.yaml"],  # noqa: ERA001
+        PROTOCOL_FILES["orders.yaml"],
+        PROTOCOL_FILES["positions.yaml"],
+        PROTOCOL_FILES["spot_asset.yaml"],
+        PROTOCOL_FILES["tickers.yaml"],
+    ],
+)
+def test_scaffold_protocol(module_scoped_dummy_agent_tim, protocol_spec: Path):
+    """Test `adev scaffold protocol` command."""
+
+    assert module_scoped_dummy_agent_tim
+
+    protocol = read_protocol_spec(protocol_spec)
+    repo_root = protodantic.get_repo_root()
+    packages_dir = repo_root / "packages"
+    protocol_outpath = packages_dir / protocol.metadata.author / "protocols" / protocol.metadata.name
+
+    if protocol_outpath.exists():
+        msg = f"Protocol already exists in dummy_agent_tim: {protocol_outpath}"
+        raise ValueError(msg)
+
+    # Point PYTHONPATH to the temporary project root so generated modules are discoverable
+    env = os.environ.copy()
+    env["PYTHONPATH"] = str(repo_root)
+
+    command = ["adev", "-v", "scaffold", "protocol", str(protocol_spec)]
+    result = subprocess.run(command, env=env, check=False, text=True, capture_output=True)
+    if result.returncode != 0:
+        msg = f"Protocol scaffolding failed: {result.stderr}"
+        raise ValueError(msg)
+
+    assert protocol_outpath.exists()
+
+    test_dir = protocol_outpath / "tests"
+    command = ["pytest", str(test_dir), "-vv", "-s", "--tb=long", "-p", "no:warnings"]
+    result = subprocess.run(
+        command,
+        env=env,
+        check=False,
+        text=True,
+        capture_output=True,
+    )
+
+    assert result.returncode == 0, f"Failed pytest on generated protocol: {result.stderr}"
```

Comment:
before, i passed the `env` with custom `PYTHONPATH` only to the last `subprocess` call and tests passed. However, no code was generated in the proper location (!) as I didn't pass the custom `env` to the `adev -v scaffold protocol` call. This resulted in a situation where no tests were ran in the subprocess, causing this test to pass without actually running any tests. We should find a better way to assert this, i.e. by checking output of the subprocess call and comparing that reported output (parsing stdout) to expected result. For now, tho, this `assert protocol_outpath.exists()` ensures that the expected outpath was generated (which fails when one does not pass the `env`)

Reply:
We have the subprocess task thing for this quite extensively used throughout the repo

Reply:
Which gives you access to all the env var stuff etc and provides the abstraction